From 2d1d8708fa156daed21bb91a0f8f69b917f3e587 Mon Sep 17 00:00:00 2001
From: Vachounet <vachounet@live.fr>
Date: Mon, 26 Oct 2020 17:05:18 +0100
Subject: [PATCH 2/4] Trebuchet: Move clear all button to actions view

[AndyCGYan: Instead of removing the original, hide it in a simple way]

Change-Id: I7a0b2729c163169f7606ad4644870d07d5165e8e
---
 .../res/drawable/ic_recents_clear_all.xml      | 18 ++++++++++++++++++
 .../res/layout/overview_actions_container.xml  | 17 ++++++++++++++++-
 .../quickstep/fallback/RecentsState.java       |  2 +-
 .../quickstep/views/LauncherRecentsView.java   |  3 +--
 .../android/quickstep/views/RecentsView.java   |  6 +++++-
 5 files changed, 41 insertions(+), 5 deletions(-)
 create mode 100644 quickstep/res/drawable/ic_recents_clear_all.xml

diff --git a/quickstep/res/drawable/ic_recents_clear_all.xml b/quickstep/res/drawable/ic_recents_clear_all.xml
new file mode 100644
index 0000000..2ad0bb8
--- /dev/null
+++ b/quickstep/res/drawable/ic_recents_clear_all.xml
@@ -0,0 +1,18 @@
+<!-- Copyright (C) 2020 The Android Open Source Project
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+          http://www.apache.org/licenses/LICENSE-2.0
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:height="24dp"
+    android:width="24dp"
+    android:viewportWidth="24"
+    android:viewportHeight="24">
+    <path android:fillColor="#FF000000" android:pathData="M5,13H19V11H5M3,17H17V15H3M7,7V9H21V7" />
+</vector>
diff --git a/quickstep/res/layout/overview_actions_container.xml b/quickstep/res/layout/overview_actions_container.xml
index fcd2e54..9d64455 100644
--- a/quickstep/res/layout/overview_actions_container.xml
+++ b/quickstep/res/layout/overview_actions_container.xml
@@ -35,6 +35,21 @@
             android:text="@string/action_screenshot"
             android:theme="@style/ThemeControlHighlightWorkspaceColor" />
 
+        <Space
+            android:id="@+id/action_clear_all_space"
+            android:layout_width="@dimen/overview_actions_button_spacing"
+            android:layout_height="1dp" />
+
+        <Button
+            android:id="@+id/action_clear_all"
+            style="@style/OverviewActionButton"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:drawableStart="@drawable/ic_recents_clear_all"
+            android:text="@string/recents_clear_all"
+            android:theme="@style/ThemeControlHighlightWorkspaceColor"
+            />
+
         <Button
             android:id="@+id/action_split"
             style="@style/OverviewActionButton"
@@ -59,4 +74,4 @@
         android:layout_gravity="bottom|center_horizontal"
         android:visibility="gone" />
 
-</com.android.quickstep.views.OverviewActionsView>
\ No newline at end of file
+</com.android.quickstep.views.OverviewActionsView>
diff --git a/quickstep/src/com/android/quickstep/fallback/RecentsState.java b/quickstep/src/com/android/quickstep/fallback/RecentsState.java
index 77aac41..358f36d 100644
--- a/quickstep/src/com/android/quickstep/fallback/RecentsState.java
+++ b/quickstep/src/com/android/quickstep/fallback/RecentsState.java
@@ -140,7 +140,7 @@ public class RecentsState implements BaseState<RecentsState> {
      * For this state, whether clear all button should be shown.
      */
     public boolean hasClearAllButton() {
-        return hasFlag(FLAG_CLEAR_ALL_BUTTON);
+        return false;
     }
 
     /**
diff --git a/quickstep/src/com/android/quickstep/views/LauncherRecentsView.java b/quickstep/src/com/android/quickstep/views/LauncherRecentsView.java
index b86d0c8..52ad63b 100644
--- a/quickstep/src/com/android/quickstep/views/LauncherRecentsView.java
+++ b/quickstep/src/com/android/quickstep/views/LauncherRecentsView.java
@@ -208,8 +208,7 @@ public class LauncherRecentsView extends RecentsView<QuickstepLauncher, Launcher
         super.setOverviewStateEnabled(enabled);
         if (enabled) {
             LauncherState state = getStateManager().getState();
-            boolean hasClearAllButton = (state.getVisibleElements(mContainer)
-                    & CLEAR_ALL_BUTTON) != 0;
+            boolean hasClearAllButton = false;
             boolean hasAddDeskButton = (state.getVisibleElements(mContainer)
                     & ADD_DESK_BUTTON) != 0;
             setDisallowScrollToClearAll(!hasClearAllButton);
diff --git a/quickstep/src/com/android/quickstep/views/RecentsView.java b/quickstep/src/com/android/quickstep/views/RecentsView.java
index 3e4b953..f5645ac 100644
--- a/quickstep/src/com/android/quickstep/views/RecentsView.java
+++ b/quickstep/src/com/android/quickstep/views/RecentsView.java
@@ -127,6 +127,7 @@ import android.view.ViewTreeObserver.OnScrollChangedListener;
 import android.view.accessibility.AccessibilityEvent;
 import android.view.accessibility.AccessibilityNodeInfo;
 import android.view.animation.Interpolator;
+import android.widget.Button;
 import android.widget.ListView;
 import android.widget.OverScroller;
 import android.widget.Toast;
@@ -556,6 +557,7 @@ public abstract class RecentsView<
     private final RecentsModel mModel;
     private final int mSplitPlaceholderSize;
     private final int mSplitPlaceholderInset;
+    private Button mActionClearAllButton;
     private final ClearAllButton mClearAllButton;
     @Nullable
     private AddDesktopButton mAddDesktopButton = null;
@@ -1213,6 +1215,8 @@ public abstract class RecentsView<
         mActionsView.updateFor3pLauncher(!supportsAppPairs());
         mSplitSelectStateController = splitController;
         mDesktopRecentsTransitionController = desktopRecentsTransitionController;
+        mActionClearAllButton = (Button) mActionsView.findViewById(R.id.action_clear_all);
+        mActionClearAllButton.setOnClickListener(this::dismissAllTasks);
     }
 
     public SplitSelectStateController getSplitSelectController() {
@@ -1688,7 +1692,7 @@ public abstract class RecentsView<
      * button fully visible, center page is Clear All button.
      */
     public boolean isClearAllHidden() {
-        return mClearAllButton.getAlpha() != 1f;
+        return true;
     }
 
     @Override
-- 
2.51.0

